### PR TITLE
Add functions to export feature data to CSV and JSON formats

### PR DESCRIPTION
**Pull request by ilkilic**
We have also eased up on type checking. Now, the system will try to convert the value to the correct type, such as int to float.

## Checklist:
- [x] Unit tests are added to cover the changes (skip if not applicable).
- [x] The changes are mentioned in the documentation (skip if not applicable).
- [ ] CHANGELOG file is updated (skip if not applicable).

### CODE REVIEW

File: efel/settings.py [hunk 1]
```diff
@@ -116,9 +116,12 @@ def set_setting(self,
         if hasattr(self, setting_name):
             expected_types = {f.name: f.type for f in fields(self)}
             expected_type = expected_types.get(setting_name)
-            if expected_type and not isinstance(new_value, expected_type):
-                raise ValueError(f"Invalid value for setting '{setting_name}'. "
-                                 f"Expected type: {expected_type.__name__}.")
+            if expected_type is not None:
+                if expected_type == float and isinstance(new_value, int):
```
Comment (AurelienJaquier) [inline efel/settings.py:120]: Are we sure that we would only have int to float case here? Maybe we could be more general by having a try/except in which we just try to convert the variable

File: tests/test_io.py [hunk 2]
```diff
@@ -294,3 +294,66 @@ def test_load_neo_file_nwb():
...
+    if os.path.exists(filename):
+        os.remove(filename)
+
+
+def test_save_feature_to_json(json_filename):
+    """Test saving of the features to JSON file."""
+    import efel
+    import os
+    import json
+    feature_values = {
```
Comment (AurelienJaquier) [inline tests/test_io.py:320]: Are we sure that this is exactly what efel would return? No numpy stuff?

Comment (AurelienJaquier) [reply tests/test_io.py:320]: Also would be interesting to have some null/None values in it (not sure which one to sure exactly, use whatever efel is using)

File: tests/test_io.py [hunk 3]
```diff
@@ -294,3 +294,66 @@ def test_load_neo_file_nwb():
             for trace in traces:
                 assert trace['stim_start'] == [250]
                 assert trace['stim_end'] == [600]
+
+
+@pytest.fixture
+def json_filename():
+    filename = 'tmp.json'
```
Comment (AurelienJaquier) [inline tests/test_io.py:301]: Could you add those paths to the .gitignore?

File: efel/io.py [hunk 4]
```diff
@@ -207,3 +209,23 @@ def load_neo_file(file_name: str, stim_start=None, stim_end=None, **kwargs) -> l
         efel_blocks.append(efel_segments)
 
     return efel_blocks
+
+
+def save_feature_to_json(feature_values, filename):
+    """Save feature values as a JSON file."""
+    with open(filename, 'w') as f:
+        json.dump(feature_values, f, default=lambda x: int(x)
+                  if isinstance(x, np.int64) else x)
```
Comment (AurelienJaquier) [inline efel/io.py:218]: Are we sure that we only encounter the np.int64 case? If so, this is ok. If not, you can have a look at how we dealt with that in EModelRunner if you need some inspiration: https://github.com/BlueBrain/EModelRunner/blob/main/emodelrunner/json_utilities.py

File: tests/test_settings.py [hunk 5]
```diff
@@ -52,6 +52,29 @@ def test_set_setting_invalid_type():
...
+    with pytest.raises(ValueError):
+        settings.set_setting("voltage_base_mode", 1)
+
+
+def test_set_setting_invalid_type3():
+    """Test that the set_setting method raises a ValueError
+    when given an invalid type."""
+    settings = Settings()
+    with pytest.raises(ValueError):
+        settings.set_setting("max_spike_skip", 1.0)
```
Comment (AurelienJaquier) [inline tests/test_settings.py:68]: Here, we could just try to convert that into an int instead of raising an error? Might be tricky for non integer float edge case, e.g. 1.1

**Review by AurelienJaquier — COMMENTED** [review]
I have a few comments :3

File: efel/settings.py (hunk 1 above)
Comment (ilkilic) [reply efel/settings.py:120]: Yes I thought about this but then it will for example allow boolean values for int features, do we want that?

Comment (AurelienJaquier) [reply efel/settings.py:120]: Indeed, allowing boolean to int might be a bit of a stretch. But I think that we should at least allow also **float to int** (if the value does not change), that might happen a lot, and **int to bool**, we have this in multiple legacy scripts. Also we might have other edge cases, like numpy.int64 that are not taken care of here and that we should allow.
I think it is better to allow too much (e.g. bool to int) with a try / except, than making efel crash on a lot of edge cases we did not anticipate. We could have an additional check afterwards (e.g. `assert value_before == value_after_conversion`) to be sure that we are not screwing the values (e.g. float to int case with `int(1.1)` )
What do you think?

Comment (ilkilic) [reply efel/settings.py:120]: I agree, I will make the changes, we could even provide a warning message to the user when we make a conversion

File: tests/test_io.py (hunk 3 above)
Comment (ilkilic) [reply tests/test_io.py:301]: I think the fixture will take care of deleting the file once the test is done, no?

Comment (AurelienJaquier) [reply tests/test_io.py:301]: Oh sorry yes. It's ok then

File: tests/test_io.py (hunk 2 above)
Comment (ilkilic) [reply tests/test_io.py]: Good point, I have switched to using the output from `get_feature_values` instead and tested with some more traces that have some or all None feature values

File: efel/io.py (hunk 4 above)
Comment (ilkilic) [reply efel/io.py]: You are right, we need to consider additional cases too. Thanks for the example, I have implemented something similar

File: efel/settings.py (hunk 1 above)
Comment (AurelienJaquier) [reply efel/settings.py]: Nice change! About the message above, I don't think we have to make a warning every time we do a conversion.
But when we change the value (e.g. input=1.1, int(input) -> 1), then I think we should either make a warning or raise an error, given that the input is far from what was expected. What do you think?

Comment (ilkilic) [reply efel/settings.py]: Indeed, after some consideration I have opted for debug statements. I agree, it seems prudent to make a warning in that case, I will add it

Comment (AurelienJaquier) [reply efel/settings.py]: I don't think we should always make a warning for float -> int. When the value does not change, like converting 1.0 to 1, I don't think we should throw a warning. It seems like something that is expected and would occur often.

Comment (ilkilic) [reply efel/settings.py]: good point, I have changed it so it raises a warning only when truncation happens

**Review by AurelienJaquier — APPROVED** [review]
Looks great, thank you Ilkan!

**Review by anilbey — APPROVED** [review]
